### DISCUSSION TITLE
Questions about meshing on a irregular shape(MegaPower, Special Purpose Nuclear, SPR)

### DISCUSSION THREAD

**Discussion by Wang-Yihu** [idaholab/moose#29076] [Q&A Meshing] [answered]
Question: How to mesh with MOOSE in this irregular shape?(It can be divided into a regular hexagon in the middle and six trimmed irregular hexagon)
![image](https://github.com/user-attachments/assets/6b04d5ce-1148-40e6-ab20-b8e4f29ca3ac)
Recent days, I'm trying to mesh on MegaPower(https://www.osti.gov/biblio/1410224). Its shape is
![image](https://github.com/user-attachments/assets/135b0a0e-65b0-4df9-9fce-6808dda752b5)
It can be seen as the duplication for the small units of the first figure in most areas.
I'm trying to mesh with MOOSE after I see this. It seems that PolygonConcentricCircleMeshGenerator, PatternedHexMeshGenerator, and  HexagonMeshTrimmer can solve my problem plausibly. 
https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial04_meshing/step06_common_ops.html
However, there is a big difference between the example of this web page and my object. In the model in the first figure, in the corner region, the hexagon is not a regular hexagon, that is, the sides of the hexagon are not equal, which I've marked in the first figure.
So can MOOSE mesh can solve this problem?

P.S, this can be solved by struct block mesh. I finish this mesh by ANSYS Workbench Mesh and ICEM CFD.
![image](https://github.com/user-attachments/assets/4f11f2bc-c43e-47c2-9b2d-d38770d370e2)

**Comment by GiudGiud** [comment]
> I'm trying to mesh with MOOSE after I see this. It seems that PolygonConcentricCircleMeshGenerator, PatternedHexMeshGenerator, and HexagonMeshTrimmer can solve my problem plausibly.

Yes this is exactly what you need to do for the very first shape. Did you get it working?
THe HexagonMeshTrimmer could be replaced by 6 CutMeshByPlane generator if it does not perform as expected.

Are you trying to mesh the second one too?

**Reply by Wang-Yihu** [reply]
Dear Giud, thanks for your reply! However, maybe my words are not clear. Now I draw the geometry again:
![image](https://github.com/user-attachments/assets/8daf7ac5-a9d1-4add-b2b5-00af3da77568)
And I put the figure of the example in url https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial04_meshing/step06_common_ops.html
with PolygonConcentricCircleMeshGenerator, PatternedHexMeshGenerator, and HexagonMeshTrimmer.
![image](https://github.com/user-attachments/assets/f95db4f3-2e00-4ef2-9501-ef09488a0ef2)
You may have noticed that there is a very big difference between these two figures: All the hexagons in the MOOSE example(the second figure) are regular hexagons. However, in my example(the first figure), only the center hexagon is regular hexagon. The other six trimmed hexagon are not regular hexagons - their edges lengths are not equal - I have marked their values in the figure.
It seems that the class PolygonConcentricCircleMeshGenerator can only mesh regular hexagon, for its required parameter polygon_size is only one number... It seems that it cannot solve this problem...
I'm looking forward to your reply...

**Reply by GiudGiud** [reply]
Oh interesting. 

Ok this can be done too but it will take more work. 

For the center hexagon, no change. 
For the side hexagons, use another PCCMG with the smaller dimensions. 
Then use a FlexiblePattern generator to place all 7 hexagons, using the custom placement option (not hexagonal, not Cartesian) since your lattice is not regular. 
https://mooseframework.inl.gov/source/meshgenerators/FlexiblePatternGenerator.html
Then use the trimmer the same way since the trim dimensions are the ones for the center hexagon

**Reply by Wang-Yihu** [reply]
Dear GiudGiud, thanks for your suggestion. Now I input PolygonConcentricCircleMeshGenerator - FlexiblePatternGenerator
```
[Mesh]
final_generator = 'fpg'
  [hex_1]
    type = PolygonConcentricCircleMeshGenerator
    # General parameters
    num_sides = 6
    num_sectors_per_side = '4 4 4 4 4 4'
    polygon_size = 0.008375
    # Ring regions parameters
    ring_radii = '0.0065 0.0070 0.007875'
    ring_intervals = '2 2 2'
    ring_block_ids = '10 15 20 25'
    ring_block_names = 'vapor_tri vapor wick clad'
    preserve_volumes = on
    # Background region parameters
    background_intervals = 2
    background_block_ids = 30
    background_block_names = background
  []
  [hex_2]
    type = PolygonConcentricCircleMeshGenerator
    num_sides = 6
    num_sectors_per_side = '4 4 4 4 4 4'
    polygon_size = 0.007625
    ring_radii = '0.00706 0.007125'
    ring_intervals = '2 1'
    ring_block_ids = '100 150 200'
    ring_block_names = 'fuel_tri fuel gap'
    preserve_volumes = on
    background_intervals = 2
    background_block_ids = 250
    background_block_names = background
  []
  #The following error occurred in the object "pattern_assm", 
  #of type "PatternedHexMeshGenerator".
  #In PatternedHexMeshGenerator pattern_assm: 
  #pitch metadata values of all input mesh generators must be identical.
#  [pattern_assm_error]
#    type = PatternedHexMeshGenerator
#    inputs = 'hex_1 hex_2'
#    pattern =  '1 1;
#               1 0 1;
#                1 1'
#    hexagon_size = 0.024
#    background_intervals = 2
#    background_block_id = 1000
#    background_block_name = background
#  []
  [fpg]
    type = FlexiblePatternGenerator
    inputs = 'hex_1 hex_2'
    boundary_type = HEXAGON
    boundary_sectors = 10
    boundary_size = 0.048
    hex_patterns = '1 1;
                   1 0 1;
                    1 1'
    hex_pitches = 0.016
  []
  #The input mesh is not compatible with HexagonMeshTrimmer 
  # because the trimmability mesh meta data are absent.
#  [hmt_err]
#    type = HexagonMeshTrimmer
#    input = fpg
#    trim_peripheral_region = '1 1 1 1 1 1'
#  []
[]
```
It will return (I mark trimming line I want with yellow )
![image](https://github.com/user-attachments/assets/16d2cd9c-fe5c-4cf1-8c8a-1d96ece2c96a)
However, HexagonMeshTrimmer does not work in this problem. It seems that it can only handle mesh from  PatternedHexMeshGenerator, not FlexiblePatternGenerator in my problem...

**Reply by GiudGiud** [reply]
Ok so that s ok. The mesh trimmer is just 6 CutMeshByPlaneGenerator

**Reply by GiudGiud** [reply]
You can replace it by chaining these plane cutting operations. The CutMeshByPlaneGenerator should be preferred over the PlaneDeletion Generator because it will cut the boundary without creating dents in the mesh

**Reply by Wang-Yihu** [reply, marked as answer]
In fact, CutMeshByPlaneGenerator cannot be applied for my 2D problem, it can only be applied for 3D problem... I use XYMeshLineCutter.
Here is my *.i file:
```
[Mesh]
final_generator =  'rename'
  [hex_1]
    type = PolygonConcentricCircleMeshGenerator
    # General parameters
    num_sides = 6
    num_sectors_per_side = '4 4 4 4 4 4'
    polygon_size = 0.008375
    # Ring regions parameters
    ring_radii = '0.0065 0.0070 0.007875'
    ring_intervals = '2 2 2'
    ring_block_ids = '10 15 20 25'
    ring_block_names = 'vapor_tri vapor wick clad'
    preserve_volumes = on
    # Background region parameters
    background_intervals = 2
    background_block_ids = 30
    background_block_names = background
  []
  [hex_2]
    type = PolygonConcentricCircleMeshGenerator
    num_sides = 6
    num_sectors_per_side = '4 4 4 4 4 4'
#    polygon_size = 0.007625
    polygon_size = 0.007250
    ring_radii = '0.00706 0.007125'
    ring_intervals = '2 1'
    ring_block_ids = '100 150 200'
    ring_block_names = 'fuel_tri fuel gap'
    preserve_volumes = on
    background_intervals = 2
    background_block_ids = 250
    background_block_names = background
  []
  #The following error occurred in the object "pattern_assm", 
  #of type "PatternedHexMeshGenerator".
  #In PatternedHexMeshGenerator pattern_assm: 
  #pitch metadata values of all input mesh generators must be identical.
  [pattern_assm_error]
    type = PatternedHexMeshGenerator
    inputs = 'hex_1 hex_2'
    pattern =  '1 1;
               1 0 1;
                1 1'
    hexagon_size = 0.024
    background_intervals = 2
    background_block_id = 1000
    background_block_name = background
  []
  [fpg]
    type = FlexiblePatternGenerator
    inputs = 'hex_1 hex_2'
    boundary_type = HEXAGON
    boundary_sectors = 10
    boundary_size = 0.048
    hex_patterns = '1 1;
                   1 0 1;
                    1 1'
    hex_pitches = 0.016
    background_subdomain_id = 1
    background_subdomain_name = monolith_quad
  []
  #The input mesh is not compatible with HexagonMeshTrimmer 
  # because the trimmability mesh meta data are absent.
  [hmt_err]
    type = HexagonMeshTrimmer
    input = rename
    trim_peripheral_region = '1 1 1 1 1 1'
  []
  #Only 3D meshes are supported. 
  #Use XYMeshLineCutter for 2D meshes. 
  #Mixed dimensional meshes are not supported at the moment.
  [cmbpg_err]
    type = CutMeshByPlaneGenerator
    input = fpg
    plane_point = '0.0 0.013856 0.0'
    plane_normal = '0.0 1.0 0.0'
  []
  [xymlc_1]
    type = XYMeshLineCutter
    input = fpg
    cut_line_params = '-1.732 1.0 -0.027712'
    new_boundary_id = 1111
    tri_elem_subdomain_name_suffix = 111
  []
  [xymlc_2]
    type = XYMeshLineCutter
    input = xymlc_1
    cut_line_params = '0.0 1.0 -0.013856'
    new_boundary_id = 2222
    tri_elem_subdomain_name_suffix = 222
  []
  [xymlc_3]
    type = XYMeshLineCutter
    input = xymlc_2
    cut_line_params = '1.732 1.0 -0.027712'
    new_boundary_id = 3333
    tri_elem_subdomain_name_suffix = 333
  []
  [xymlc_4]
    type = XYMeshLineCutter
    input = xymlc_3
    cut_line_params = '1.732 -1.0 -0.027712'
    new_boundary_id = 4444
    tri_elem_subdomain_name_suffix = 444
  []
  [xymlc_5]
    type = XYMeshLineCutter
    input = xymlc_4
    cut_line_params = '0.0 -1.0 -0.013856'
    new_boundary_id = 5555
    tri_elem_subdomain_name_suffix = 555
  []
  [xymlc_6]
    type = XYMeshLineCutter
    input = xymlc_5
    cut_line_params = '-1.732 -1.0 -0.027712'
    new_boundary_id = 6666
    tri_elem_subdomain_name_suffix = 666
  []
  #It seems that all classes above cannot name block easily...
  [rename]
    type = RenameBlockGenerator 
    input = xymlc_6
    old_block = '0            10        15         20   25   30            
                 100      150       200 250'
    new_block = 'monolith_tri vapor_tri vapor_quad wick clad monolith_quad 
                 fuel_tri fuel_quad gap monolith_quad'
  []
[]
```
And here is my result:
![image](https://github.com/user-attachments/assets/d28751dc-a44e-4fb6-8bc3-d96cfb76b31c)
And you maybe have noticed that there is a strange block named_with 150_222 with very very small volume. It might be generated by XYMeshLineCutter with creating dents in the mesh as you said...
So how to solve this defect? Just use BlockDeletionGenerator to delete it?
P.S. the class CutMeshByPlaneGenerator is not in my MOOSE version . I have to download CutMeshByPlaneGenerator.C/h and MooseMeshElementConversionUtils.C/h.

**Reply by GiudGiud** [reply]
I would either delete it or try to play with the directions (introducing a slight error in the direction that removes the block) for XY mesh line cutting to remove it. 

It's due to numerical precision errors in the cutting process. Deleting it should be safe

**Reply by Wang-Yihu** [reply]
Ok, very glad to see your reply! I might spend weeks or even months to solve it without your precious help!
But I still want to generate this mesh by block structured method. This geometry is so regular (even with small defect). Here I show my block structured layout and mesh by ANSYS Workbench mesh
![image](https://github.com/user-attachments/assets/ffb8d550-8915-4b09-a7c8-f2469ffb4c19)
Do you have some ideas about realizing this type mesh by MOOSE?

**Reply by GiudGiud** [reply]
Hello

We don't have this block structured algorithm in moose. Do you just mean you want to use quadrilateral elements?

Why don't you use your ansys mesh? If you look at the FileMeshGenerator documentation you can see all the formats we support

Guillaume

**Reply by GiudGiud** [reply]
If you want a quad mesh it s possible but it will take some patience. You ll have to figure out if you can the peripheral pins when they use quadrilateral center elements (a parameter of the PCCMG). If that works we can figure out how to mesh the background region with quads

**Reply by Wang-Yihu** [reply]
Haha, it will take us too many steps to realize the mesh of the full reactor core (consists of hundreds of small unit we discussed) with ANSYS. So i'm trying to realize it with MOOSE input files...

**Reply by Wang-Yihu** [reply]
Oh, sir. Maybe i did not make it clear. What i want mostly is like this: (not for circle region)
![image](https://github.com/user-attachments/assets/faf493f1-e893-4f68-8249-331e57c63dc7)
If I know every quadrilateral block coordinate, can i use moose to realize structured mesh like ANSYS ICEM CFD?

**Reply by GiudGiud** [reply]
Can you just delete the pins from the mesh you just generated with moose?

If you know all the node coordinates you can generate elements one by one yes. There are probably better tools than moose for that though, like creating a msh gmsh file from Python 
In moose I think it s the ElementGenerator

**Reply by Wang-Yihu** [reply]
Ok, thanks for your reply. I'll try GMesh as you suggested in the future. Now I mark my reply with *.i file as the answer.
